Implement partial application and constant folding.

sellout opened the pull request:
This adds a number of additional “simplifications” to the optimizer. A
few are handled directly, but for Invoke, a new parameter has been added
to Func. It’s done this way because the constant folding should handled
in the Typer function, and it makes sense to keep the other
simplifications near there.

There were a few bugs fixed along the way:

* MongoDbPlanner couldn’t handle constant arrays for sort directions,
* `Take` and `Drop` expected the wrong number of arguments, and
* `Substring` handled indexes inconsistently (see #795).

Many constant folds are currently duplicated in the simplifiers. This is
because I had already implemnted some before I realized the typers
should handle that. I left them in because some of them aren’t yet
reflected in the typers and because we’re not yet applying the folds
from the typers. (Doing that will require adding source locations and
moving the typechecker to the LogicalPlan rather than the SQL AST.)

Also, this will cause the example in #761 to not fail (because it elides the constant filter), but it doesn’t fix the underlying issue.

jdegoes commented:
Do you think is a good idea to add all this machinery when some of it is duplicated and most of it will go away in the not-to-distant future? Also technically the existing folding in typers could be utilized in the SQL AST, mainly by treating Type constants as literals when constructing the logical plan for them.

sellout commented:
I think the machinery is useful regardless, as there are some cases that aren’t handled by constant folding, and keeping all the optimizations for each function together seems like the right thing.

Applying the constant folding when converting to LP is a good idea – I wanted to do it during typechecking, and that wasn’t an option. I can do that and make sure all the constant folding cases in simplify are reflected in the typer (there are at least a few that aren’t yet).

mossprescott commented:
Does this fix #795, or just touch on it? I guess it just fixes the un-recorded bugs you listed?

I'm also concerned about the amount of code here and how the heck we're going to cover the code we write in this layer in the future. And how we're going to know that it's consistent with constant-folding in the typers, and with evaluation in the pipeline, and in map-reduce. You've essentially added another evaluation step; granted, we probably need it sooner or later. I guess the question is whether we can unify this step with constant folding in the typers, say if we moved that step to the logical plan.

Feels like you've just shined a light on a place where the plan is vague.

sellout commented:
No, I added #795 as a result of this. Found that we were handling it inconsistently, fixed it to be internally consistent (although not consistent with Postgres (ANSI SQL?) semantics) and opened the issue.

As far as this being consistent with constant-folding, it should actually be orthogonal to constant-folding (I’ll remove the duplication before merge). We’re supposedly already applying constant-folding in `compile0`, but at least in some cases it doesn’t work. EG, this PR simplifies `select * from zips where false` correctly, but without this it doesn’t (even though that fold exists in the `Filter` typer).

This PR probably fixes #777, and as I mentioned there, it’s often harder to identify available folds in the SQL AST, and that’s likely why we miss the `Filter` case above. But probably best to not duplicate that in the simplifier, and just wait until typechecking can happen on the LP for those cases to work.

sellout commented:
With this update, there are now many more `noSimplification` cases. I still think it’s useful to keep the simplifiers near the typers, but also an option at this point to move them all into `Optimizer.simplify`.

mossprescott commented:
One bad comment and a couple of questions/suggestions. Otherwise I see no new issues.

Not sure where we landed on the question of whether this a a good direction to go with typers.

sellout commented:
As far as whether the typer approach is good or not, that’s an easy change. I like it, but am happy to put all the simplifiers in `simplify` if others prefer it. It could also be changed later on without much work, though.

mossprescott commented on core/src/main/scala/slamdata/engine/functions.scala:
```diff
@@ -54,8 +58,9 @@ trait FuncInstances {
 object Func extends FuncInstances {
-  type Typer   = List[Type] => ValidationNel[SemanticError, Type]
-  type Untyper = Type => ValidationNel[SemanticError, List[Type]]
+  type Simplifier = List[Term[LogicalPlan]] => Func => Term[LogicalPlan]
```
Now that I'm looking at this again, this type is raising a lot of questions for me.

Why is this function total? That means either: we're calling it and then comparing the result with the original term (so, it's partial, just in a sneaky way), or we're just calling it exactly once and not attempting to simplify the result, but simplification/partial evaluation typically begets more simplification, so don't we want to do that? I guess I would expect it to produce an `Option`, just to make the partiality explicit.

Why the weird curried type? As far as I can tell, no one actually needs the `Func` parameter, except the two combinators. It seems awkward, and then of course you want to call it in a simple way, so we end up with another `0` method. There should be a simpler way to express this, maybe using `this` to refer to the enclosing `Func`?

sellout replied:
You’re right – it should be Option. It doesn’t currently re-simplify the result. I’ll fix that.

It doesn’t have to be curried, that’s just how I think. I’ll fix that, too. But we do need the `Func` parameter, because a `Simplifier` is passed to the `Func` constructor, so there is no enclosing `Func` when it’s created. But it is eventually passed `this` when it’s called.

sellout replied:
Actually, it only needed the `Func` to build up the `InvokeF` in the case of there not being a simplification. Now that we return `Option`, it’s no longer needed, so it’s gone 😃

mossprescott replied:
Nice!

mossprescott commented:
Hmm. 

You've also made the std library functions aware of LogicalPlan, which is a new dependency. I don't think that's terrible, since LP is pretty abstract, but it was simpler when the funcs only knew about types (and Data).

I'm also not happy about the duplication. If the simplifiers only ever match on `Constant(...)`, then shouldn't you be able to apply the typer to get the same result? Can we write a typer-based simplifier that does that, and then only have separate implementations when we're looking for patterns that don't involve constants? What would those be?

This also relates to another issue I need to look at soon ("simplifying" expressions so they can be turned into selectors)...

Maybe we should talk this over in real time?

sellout commented:
You can only apply the typer in the case where the _result_ is a constant.

So, “Can we write a typer-based simplifier that does that, and then only have separate implementations when we're looking for patterns that don't involve constants? What would those be?”

That’s what we’re doing now – the typer handles constant folding, and the simplifier handles cases that result in a non-constant node.

Yeah, happy to talk in real-time. Monday, 10:00? 😃

mossprescott commented:
Ah, I see. Maybe you can remind me on Monday what doing constant folding in the typer buys us. Talk to you then!
